File: Patrones Estructurales/Decorator/decorator.py
from abc import ABC, abstractmethod

# 1. Component
class Cafe(ABC):
    @abstractmethod
    def costo(self) -> int:
        pass
    
    @abstractmethod
    def descripcion(self) -> str:
        pass

# Componente Concreto (Base)
class CafeSimple(Cafe):
    def costo(self) -> int:
        return 5
    def descripcion(self) -> str:
        return "Café simple"

# 2. Decorator Base
class DecoradorCafe(Cafe):
    def __init__(self, cafe: Cafe):
        self._cafe = cafe

    def costo(self) -> int:
        return self._cafe.costo()
    
    def descripcion(self) -> str:
        return self._cafe.descripcion()

# 3. Concrete Decorators
class ConLeche(DecoradorCafe):
    def costo(self) -> int:
        return self._cafe.costo() + 2
    
    def descripcion(self) -> str:
        return self._cafe.descripcion() + ", con Leche"

# Uso
mi_cafe = CafeSimple()
mi_cafe = ConLeche(mi_cafe) # Envolvemos

print(f"{mi_cafe.descripcion()} cuesta ${mi_cafe.costo()}")

File: Patrones Estructurales/Decorator/decorator.js
// 1. Component (Interfaz común)
class Cafe {
    costo() { return 5; } // Costo base
    descripcion() { return "Café simple"; }
}

// 2. Decorator (Base - Mantiene la referencia)
class DecoradorCafe extends Cafe {
    constructor(cafe) {
        super();
        this.cafe = cafe; // Referencia al objeto envuelto
    }

    costo() { return this.cafe.costo(); }
    descripcion() { return this.cafe.descripcion(); }
}

// 3. Concrete Decorators (Los agregados)
class ConLeche extends DecoradorCafe {
    costo() {
        return this.cafe.costo() + 2; // Suma su precio
    }
    descripcion() {
        return this.cafe.descripcion() + ", con Leche";
    }
}

class ConAzucar extends DecoradorCafe {
    costo() {
        return this.cafe.costo() + 1;
    }
    descripcion() {
        return this.cafe.descripcion() + ", con Azúcar";
    }
}

// Uso
let miCafe = new Cafe(); // $5
miCafe = new ConLeche(miCafe); // $7 (5+2)
miCafe = new ConAzucar(miCafe); // $8 (7+1)

console.log(`${miCafe.descripcion()} cuesta $${miCafe.costo()}`);
// Salida: Café simple, con Leche, con Azúcar cuesta $8